File: rosen/test/2d_graphics/drawing_ndk/cpp/interface/matrix_test.cpp
/*
 * Copyright (c) 2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "matrix_test.h"

#include <native_drawing/drawing_brush.h>
#include <native_drawing/drawing_color.h>
#include <native_drawing/drawing_filter.h>
#include <native_drawing/drawing_image.h>
#include <native_drawing/drawing_matrix.h>
#include <native_drawing/drawing_path.h>
#include <native_drawing/drawing_path_effect.h>
#include <native_drawing/drawing_pen.h>
#include <native_drawing/drawing_point.h>
#include <native_drawing/drawing_round_rect.h>
#include <native_drawing/drawing_sampling_options.h>
#include <native_drawing/drawing_shader_effect.h>

#include "test_common.h"

#include "common/log_common.h"

void MatrixReset::OnTestPerformance(OH_Drawing_Canvas* canvas)
{
    OH_Drawing_Path* path = OH_Drawing_PathCreate();
    OH_Drawing_PathAddRect(path, 0, 0, 100, 100, OH_Drawing_PathDirection::PATH_DIRECTION_CW); // 100 矩形宽高
    OH_Drawing_Matrix* matrix = OH_Drawing_MatrixCreate();
    OH_Drawing_MatrixSetMatrix(matrix, 50, 0, 0, 0, 50, 0, 0, 0, 50); // 50 matrix值

    for (int i = 0; i < testCount_; i++) {
        OH_Drawing_MatrixReset(matrix);
    }
    OH_Drawing_PathTransform(path, matrix);
    OH_Drawing_CanvasDrawPath(canvas, path);
    OH_Drawing_MatrixDestroy(matrix);
    OH_Drawing_PathDestroy(path);
}

void MatrixConcat::OnTestPerformance(OH_Drawing_Canvas* canvas)
{
    OH_Drawing_Path* path = OH_Drawing_PathCreate();
    OH_Drawing_Matrix* matrix_a = OH_Drawing_MatrixCreate();
    OH_Drawing_Matrix* matrix_b = OH_Drawing_MatrixCreate();
    OH_Drawing_Matrix* total = OH_Drawing_MatrixCreate();
    OH_Drawing_PathAddRect(path, 0, 0, 100, 100, OH_Drawing_PathDirection::PATH_DIRECTION_CW); // 100 宽高
    OH_Drawing_MatrixSetMatrix(matrix_a, 50, 0, 0, 0, 50, 0, 0, 0, 50); // 50 matrix值
    OH_Drawing_MatrixSetMatrix(matrix_b, 60, 0, 0, 0, 60, 0, 0, 0, 60); // 60 matrix值

    for (int i = 0; i < testCount_; i++) {
        OH_Drawing_MatrixConcat(total, matrix_a, matrix_b);
    }
    OH_Drawing_PathTransform(path, total);
    OH_Drawing_CanvasDrawPath(canvas, path);
    OH_Drawing_MatrixDestroy(total);
    OH_Drawing_MatrixDestroy(matrix_a);
    OH_Drawing_MatrixDestroy(matrix_b);
    OH_Drawing_PathDestroy(path);
}

void MatrixInvert::OnTestPerformance(OH_Drawing_Canvas* canvas)
{
    OH_Drawing_Path* path = OH_Drawing_PathCreate();
    OH_Drawing_Matrix* matrix = OH_Drawing_MatrixCreate();
    OH_Drawing_Matrix* inverse = OH_Drawing_MatrixCreate();
    OH_Drawing_PathAddRect(path, 0, 0, 100, 100, OH_Drawing_PathDirection::PATH_DIRECTION_CW); // 100 宽高
    OH_Drawing_MatrixSetMatrix(matrix, 50, 0, 0, 0, 50, 0, 0, 0, 50); // 50 matrix值

    for (int i = 0; i < testCount_; i++) {
        OH_Drawing_MatrixInvert(matrix, inverse);
    }
    OH_Drawing_PathTransform(path, inverse);
    OH_Drawing_CanvasDrawPath(canvas, path);
    OH_Drawing_MatrixDestroy(matrix);
    OH_Drawing_MatrixDestroy(inverse);
    OH_Drawing_PathDestroy(path);
}

void MatrixSetPolyToPoly::OnTestPerformance(OH_Drawing_Canvas* canvas)
{
    int count = 4;
    OH_Drawing_Path* path = OH_Drawing_PathCreate();
    OH_Drawing_PathAddRect(path, 0, 0, 100, 100, OH_Drawing_PathDirection::PATH_DIRECTION_CW); // 100 宽高
    OH_Drawing_Matrix* matrix = OH_Drawing_MatrixCreate();
    OH_Drawing_Point2D src[] = { { 0, 0 }, { 100, 0 }, { 0, 100 }, { 100, 100 } }; // 0, 100 坐标值
    OH_Drawing_Point2D dst[] = { { 20, 20 }, { 20, 200 }, { 200, 20 }, { 200, 200 } }; // 20, 200 坐标值
    for (int i = 0; i < testCount_; i++) {
        OH_Drawing_MatrixSetPolyToPoly(matrix, src, dst, count);
    }
    OH_Drawing_PathTransform(path, matrix);
    OH_Drawing_CanvasDrawPath(canvas, path);
    OH_Drawing_MatrixDestroy(matrix);
    OH_Drawing_PathDestroy(path);
}
void MatrixPreRotate::OnTestPerformance(OH_Drawing_Canvas* canvas)
{
    // 性能测试:10000次4ms
    OH_Drawing_Matrix* pMatrix = OH_Drawing_MatrixCreate();
    OH_Drawing_Matrix* matrix = OH_Drawing_MatrixCreate();
    OH_Drawing_MatrixSetMatrix(matrix, 2, -1, 2, 5, -4, 5, 8, -7, 8); // 2, -1, 2, 5, -4, 5, 8, -7, 8矩阵参数
    // 将矩阵设置为矩阵左乘围绕轴心点旋转一定角度的单位矩阵后得到的矩阵
    for (int i = 0; i < testCount_; i++) {
        OH_Drawing_MatrixSetMatrix(pMatrix, 1, 2, 3, 4, 5, 6, 7, 8, 9); // 1, 2, 3, 4, 5, 6, 7, 8, 9 矩阵参数
        OH_Drawing_MatrixPreRotate(pMatrix, 90, 1, 0);                  // 90, 1矩阵角度和坐标
    };
    // 观察pMatrix转换后与matrix是否一样.来判断OH_Drawing_MatrixPreRotate接口是否调用成功
    bool result = false;
    result = OH_Drawing_MatrixIsEqual(pMatrix, matrix);
    if (result == true) {
        DRAWING_LOGI("The OH_Drawing_MatrixPreRotate interface returns a value of true,result=%{public}s",
            result ? "true" : "false");
        OH_Drawing_Rect* rect = OH_Drawing_RectCreate(0, 0, 100, 100); // 0, 0, 100, 100 创建矩形
        OH_Drawing_CanvasDrawRect(canvas, rect);
        OH_Drawing_RectDestroy(rect);
    } else {
        DRAWING_LOGI("The OH_Drawing_MatrixPreRotate interface returns a value of failed,result=%{public}s",
            result ? "true" : "false");
    }
    // 释放
    OH_Drawing_MatrixDestroy(pMatrix);
    OH_Drawing_MatrixDestroy(matrix);
}

void MatrixPostScale::OnTestPerformance(OH_Drawing_Canvas* canvas)
{
    // 性能测试:10000次27ms
    OH_Drawing_Matrix* pMatrix = OH_Drawing_MatrixCreate();
    OH_Drawing_Matrix* matrix = OH_Drawing_MatrixCreate();
    OH_Drawing_MatrixSetMatrix(matrix, -37, -11, -3, -99, -23, -23, 3, 1, 1); // -37, -11, -3, -99, -23, 3, 1矩阵参数
    // 将矩阵设置为矩阵右乘围绕轴心点按一定缩放因子缩放后的单位矩阵后得到的矩阵
    for (int i = 0; i < testCount_; i++) {
        OH_Drawing_MatrixSetMatrix(pMatrix, 2, 1, 3, 1, 2, 2, 3, 1, 1); // 2, 1, 3, 1, 2, 2, 3, 1矩阵参数
        OH_Drawing_MatrixPostScale(pMatrix, 4, 6, 5, 7);                // 4, 6, 5, 7缩放比例因子和坐标
    };
    // 观察pMatrix转换后与matrix是否一样.来判断OH_Drawing_MatrixPostScale接口是否调用成功
    bool result = false;
    result = OH_Drawing_MatrixIsEqual(pMatrix, matrix);
    if (result == true) {
        DRAWING_LOGI("The OH_Drawing_MatrixPostScale interface returns a value of true,result=%{public}s",
            result ? "true" : "false");
        OH_Drawing_Rect* rect = OH_Drawing_RectCreate(0, 0, 100, 100); // 0, 0, 100, 100 创建矩形
        OH_Drawing_CanvasDrawRect(canvas, rect);
        OH_Drawing_RectDestroy(rect);
    } else {
        DRAWING_LOGI("The OH_Drawing_MatrixPostScale interface returns a value of failed,result=%{public}s",
            result ? "true" : "false");
    }
    // 释放
    OH_Drawing_MatrixDestroy(pMatrix);
    OH_Drawing_MatrixDestroy(matrix);
}

void MatrixPostTranslate::OnTestPerformance(OH_Drawing_Canvas* canvas)
{
    // 性能测试:10000次2ms
    OH_Drawing_Matrix* pMatrix = OH_Drawing_MatrixCreate();
    OH_Drawing_Matrix* matrix = OH_Drawing_MatrixCreate();
    OH_Drawing_MatrixSetMatrix(matrix, 8, 3, 5, 13, 6, 6, 3, 1, 1); // 8, 3, 5, 13, 6, 6, 3, 1矩阵参数
    // 将矩阵设置为矩阵右乘平移一定距离后的单位矩阵后得到的矩阵.
    for (int i = 0; i < testCount_; i++) {
        OH_Drawing_MatrixSetMatrix(pMatrix, 2, 1, 3, 1, 2, 2, 3, 1, 1); // 2, 1, 3, 1, 2, 2, 3, 1矩阵参数
        OH_Drawing_MatrixPostTranslate(pMatrix, 2, 4);                  // 2, 4平移距离
    }
    // 观察pMatrix转换后与matrix是否一样.来判断OH_Drawing_MatrixPostTranslate接口是否调用成功
    bool result = false;
    result = OH_Drawing_MatrixIsEqual(pMatrix, matrix);
    if (result == true) {
        DRAWING_LOGI("The OH_Drawing_MatrixPostTranslate interface returns a value of true,result=%{public}s",
            result ? "true" : "false");
        OH_Drawing_Rect* rect = OH_Drawing_RectCreate(0, 0, 100, 100); // 0, 0, 100, 100 创建矩形
        OH_Drawing_CanvasDrawRect(canvas, rect);
        OH_Drawing_RectDestroy(rect);
    } else {
        DRAWING_LOGI("The OH_Drawing_MatrixPostTranslate interface returns a value of failed,result=%{public}s",
            result ? "true" : "false");
    }
    // 释放
    OH_Drawing_MatrixDestroy(pMatrix);
    OH_Drawing_MatrixDestroy(matrix);
}

void MatrixIsEqual::OnTestPerformance(OH_Drawing_Canvas* canvas)
{
    OH_Drawing_Matrix* matrix1 = OH_Drawing_MatrixCreate();
    OH_Drawing_Matrix* matrix2 = OH_Drawing_MatrixCreate();
    OH_Drawing_Rect* rect = OH_Drawing_RectCreate(0, 0, 100, 100); // 0, 0, 100, 100 创建矩形
    bool matrixIsEqual = false;
    for (int i = 0; i < testCount_; i++) {
        matrixIsEqual = OH_Drawing_MatrixIsEqual(matrix1, matrix2);
    }
    if (matrixIsEqual) {
        OH_Drawing_CanvasDrawRect(canvas, rect);
    }
    OH_Drawing_RectDestroy(rect);
    OH_Drawing_MatrixDestroy(matrix1);
    OH_Drawing_MatrixDestroy(matrix2);
}

File: rosen/test/2d_graphics/drawing_ets/ets/testcase/interface/matrixtest.ts
/*
 * Copyright (c) 2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import drawing from "@ohos.graphics.drawing";
import common2D from "@ohos.graphics.common2D";
import { OHRandom } from '../../utils/OHRandom';
import { TestBase, TestFunctionStyleType, StyleType } from '../../pages/testbase';

const TAG = '[BezierBench]';

function displayTestResult(canvas: drawing.Canvas, isPassed : Boolean) {
  let testResult : string = isPassed ? "PASSED" : "FAILED. Please check the console logs";
  let font : drawing.Font = new drawing.Font();
  font.setSize(50);
  let blob : drawing.TextBlob = drawing.TextBlob.makeFromString(testResult, font, drawing.TextEncoding.TEXT_ENCODING_UTF8);
  canvas.drawTextBlob(blob, 10, 100);
}

function checkMatrix(actualMatrix : drawing.Matrix, expectedValues: Array<number>): boolean {
  // returns FALSE if matrix has unexpected element value
  let failure: boolean = false;
  let fractionDigits: number = 5;
  for (let i = 0; i < expectedValues.length; i++) {
    if (actualMatrix.getValue(i).toFixed(fractionDigits) != expectedValues[i].toFixed(fractionDigits) ) {
      failure = true;
      console.log(TAG, "The matrix[" + i + "] is incorrect (actual: " + actualMatrix.getValue(i)
        + ", expected: " + expectedValues[i] + ")");
    }
  }
  return !failure;
}

export class MatrixGetValue extends TestBase {
  public constructor() {
    super();
  }

  public OnTestFunction(canvas: drawing.Canvas) {
    let expectedValues: Array<number> = [1, 0, 0, 0, 1, 0, 0, 0, 1];
    let matrix = new drawing.Matrix();

    let failure: boolean = false;
    for (let i = 0; i < expectedValues.length; i++) {
      if (matrix.getValue(i) != expectedValues[i]) {
        failure = true;
        console.log(TAG, "The matrix[" + i + "] is incorrect (actual: " + matrix.getValue(i)
          + ", expected: " + expectedValues[i] + ")");
      }
    }
    displayTestResult(canvas, !failure);
  }

  public OnTestPerformance(canvas: drawing.Canvas) {
    let matrix = new drawing.Matrix();
    for (let i = 0; i < this.testCount_; i++) {
      matrix.getValue(0);
    }
    displayTestResult(canvas, true);
  }
}

export class MatrixGetAll extends TestBase {
  public constructor() {
    super();
  }

  public OnTestFunction(canvas: drawing.Canvas) {
    let failure: boolean = false;
    let matrix = new drawing.Matrix();
    let expectedValues: Array<number> = [1, 0, 0, 0, 1, 0, 0, 0, 1];

    let actualValues = matrix.getAll();
    if (actualValues.length != expectedValues.length) {
      failure = true;
      console.log(TAG, "The length of returned array is incorrect (actual: " + actualValues.length
        + ", expected: " + expectedValues.length  + ")");
    } else {
      for (let i = 0; i < actualValues.length; i++) {
        if (actualValues[i] != expectedValues[i]) {
          failure = true;
          console.log(TAG, "The matrix[" + i + "] is incorrect (actual: " + actualValues[i]
            + ", expected: " + expectedValues[i] + ")");
        }
      }
    }
    displayTestResult(canvas, !failure);
  }

  public OnTestPerformance(canvas: drawing.Canvas) {
    let matrix = new drawing.Matrix();
    for (let i = 0; i < this.testCount_; i++) {
      let matrixBuffer = matrix.getAll();
    }
    displayTestResult(canvas, true);
  }
}

export class MatrixPostRotate extends TestBase {

  public constructor() {
    super();
  }

  public OnTestFunction(canvas: drawing.Canvas) {
    let matrix = new drawing.Matrix();
    let degree: number = 2;
    let px: number = 3;
    let py: number = 4;
    matrix.postRotate(degree, px, py);

    let c = Math.cos(degree * Math.PI / 180);
    let s = Math.sin(degree * Math.PI / 180);
    let dx = s * py + (1 - c) * px;
    let dy = -s * px + (1 - c) * py;
    let expectedValues: Array<number> = [c, -s, dx, s, c, dy, 0, 0, 1];

    displayTestResult(canvas, checkMatrix(matrix, expectedValues));
  }

  public OnTestPerformance(canvas: drawing.Canvas) {
    let matrix = new drawing.Matrix();
    for (let i = 0; i < this.testCount_; i++) {
      matrix.postRotate(2, 3, 4);
    }
    displayTestResult(canvas, true);
  }
}

export class MatrixPostTranslate extends TestBase {

  public constructor() {
    super();
  }

  public OnTestFunction(canvas: drawing.Canvas) {
    let matrix = new drawing.Matrix();
    let dx: number = 3;
    let dy: number = 4;
    matrix.postTranslate(dx, dy);

    let expectedValues: Array<number> = [1, 0, dx, 0, 1, dy, 0, 0, 1];
    displayTestResult(canvas, checkMatrix(matrix, expectedValues));
  }

  public OnTestPerformance(canvas: drawing.Canvas) {
    let matrix = new drawing.Matrix();
    for (let i = 0; i < this.testCount_; i++) {
      matrix.postTranslate(3, 4);
    }
    displayTestResult(canvas, true);
  }
}

export class MatrixReset extends TestBase {

  public constructor() {
    super();
  }

  public OnTestFunction(canvas: drawing.Canvas) {
    let expectedValues: Array<number> = [1, 0, 0, 0, 1, 0, 0, 0, 1];
    let matrix = new drawing.Matrix();
    matrix.postRotate(2, 3, 4);
    matrix.reset();
    displayTestResult(canvas, checkMatrix(matrix, expectedValues));
  }

  public OnTestPerformance(canvas: drawing.Canvas) {
    let matrix = new drawing.Matrix();
    for (let i = 0; i < this.testCount_; i++) {
      matrix.reset();
    }
    displayTestResult(canvas, true);
  }
}

export class MatrixSetPolyToPoly extends TestBase {

  public constructor() {
    super();
  }

  public OnTestFunction(canvas: drawing.Canvas) {
    let srcPoints: Array<common2D.Point> = [ { x: 200, y: 20 }, { x: 200, y: 60 } ];
    let dstPoints: Array<common2D.Point> = [ { x:0, y: 0 }, { x:0, y: 0 } ];
    let brush = new drawing.Brush();
    let color: common2D.Color = { alpha: 255, red: 0, green: 255, blue: 0 };
    let matrix = new drawing.Matrix();
    canvas.setMatrix(matrix);
    brush.setColor(color);
    canvas.attachBrush(brush);
    console.log(TAG, "matrix.setPolyToPoly1");
    let result: boolean = false;
    try {
      if (matrix.setPolyToPoly(srcPoints, dstPoints, 2)) {
        console.log(TAG, "canvas.drawLine(dstPoints");
        canvas.drawLine(dstPoints[0].x, dstPoints[0].y, dstPoints[1].x, dstPoints[1].y);
        result = true;
      }
    } catch(err) {
      console.error("Matrix.setPolyToPoly exception: ", err.name, ":", err.message, err.stack);
    }
    canvas.detachBrush();
    if (!result) {
      displayTestResult(canvas, false);
    }
  }

  public OnTestPerformance(canvas: drawing.Canvas) {
    let matrix = new drawing.Matrix();
    let srcPoints: Array<common2D.Point> = [ { x: 200, y: 20 }, { x: 200, y: 60 } ];
    let dstPoints: Array<common2D.Point> = [ { x:0, y: 0 }, { x:0, y: 0 } ];
    let result: boolean = true;
    for (let i = 0; i < this.testCount_; i++) {
      try {
        matrix.setPolyToPoly(srcPoints, dstPoints, 2);
      } catch(err) {
        console.error("Matrix.setPolyToPoly exception: ", err.name, ":", err.message, err.stack);
        result = false;
      }
    }
    displayTestResult(canvas, result);
  }
}

export class MatrixSetRectToRect extends TestBase {

  public constructor() {
    super();
  }

  public OnTestFunction(canvas: drawing.Canvas) {
    let src: common2D.Rect = { left: 100, top: 20, right: 130, bottom: 60 };
    let dst: common2D.Rect = { left: 100, top: 80, right: 130, bottom: 120 };
    let stf: drawing.ScaleToFit = drawing.ScaleToFit.FILL_SCALE_TO_FIT;
    let brush = new drawing.Brush();
    let color: common2D.Color = { alpha: 255, red: 0, green: 255, blue: 0 };
    let matrix = new drawing.Matrix();
    canvas.setMatrix(matrix);
    brush.setColor(color);
    canvas.attachBrush(brush);
    console.log(TAG, "matrix.setRectToRect1");
    if (matrix.setRectToRect(src, dst, stf)) {
        console.log(TAG, "canvas.drawRect");
        canvas.drawRect(dst);
    }
    canvas.detachBrush();
  }

  public OnTestPerformance(canvas: drawing.Canvas) {
    let matrix = new drawing.Matrix();
    let src: common2D.Rect = { left: 100, top: 20, right: 130, bottom: 60 };
    let dst: common2D.Rect = { left: 100, top: 80, right: 130, bottom: 120 };
    let stf: drawing.ScaleToFit = drawing.ScaleToFit.FILL_SCALE_TO_FIT;
    for (let i = 0; i < this.testCount_; i++) {
      matrix.setRectToRect(src, dst, stf);
    }
    displayTestResult(canvas, true);
  }
}

export class MatrixPreScale extends TestBase {

  public constructor() {
    super();
  }

  public OnTestFunction(canvas: drawing.Canvas) {
    let matrix = new drawing.Matrix();
    let sx: number = 2;
    let sy: number = 0.5;
    let px: number = 1;
    let py: number = 1;

    let dx = px - sx * px;
    let dy = py - sy * py;
    matrix.preScale(sx, sy, px, py);

    let expectedValues: Array<number> = [sx, 0, dx, 0, sy, dy, 0, 0, 1];
    displayTestResult(canvas, checkMatrix(matrix, expectedValues));
  }

  public OnTestPerformance(canvas: drawing.Canvas) {
    let matrix = new drawing.Matrix();
    for (let i = 0; i < this.testCount_; i++) {
      matrix.preScale(2, 0.5, 0, 0);
    }
    displayTestResult(canvas, true);
  }
}

export class MatrixPreTranslate extends TestBase {

  public constructor() {
    super();
  }

  public OnTestFunction(canvas: drawing.Canvas) {
    let matrix = new drawing.Matrix();
    let dx: number = 3;
    let dy: number = 4;
    matrix.preTranslate(dx, dy);

    let expectedValues: Array<number> = [1, 0, dx, 0, 1, dy, 0, 0, 1];
    displayTestResult(canvas, checkMatrix(matrix, expectedValues));
  }

  public OnTestPerformance(canvas: drawing.Canvas) {
    let matrix = new drawing.Matrix();
    for (let i = 0; i < this.testCount_; i++) {
      matrix.preTranslate(3, 4);
    }
    displayTestResult(canvas, true);
  }
}

export class MatrixPreRotate extends TestBase {

  public constructor() {
    super();
  }

  public OnTestFunction(canvas: drawing.Canvas) {
    let matrix = new drawing.Matrix();
    let degree: number = 2;
    let px: number = 3;
    let py: number = 4;
    matrix.preRotate(degree, px, py);

    let c = Math.cos(degree * Math.PI / 180);
    let s = Math.sin(degree * Math.PI / 180);
    let dx = s * py + (1 - c) * px;
    let dy = -s * px + (1 - c) * py;
    let expectedValues: Array<number> = [c, -s, dx, s, c, dy, 0, 0, 1];

    displayTestResult(canvas, checkMatrix(matrix, expectedValues));
  }

  public OnTestPerformance(canvas: drawing.Canvas) {
    let matrix = new drawing.Matrix();
    for (let i = 0; i < this.testCount_; i++) {
      matrix.postRotate(2, 3, 4);
    }
    displayTestResult(canvas, true);
  }
}

export class MatrixPostScale extends TestBase {

  public constructor() {
    super();
  }

  public OnTestFunction(canvas: drawing.Canvas) {
    let matrix = new drawing.Matrix();
    let sx: number = 2;
    let sy: number = 0.5;
    let px: number = 1;
    let py: number = 1;

    let dx = px - sx * px;
    let dy = py - sy * py;
    matrix.preScale(sx, sy, px, py);

    let expectedValues: Array<number> = [sx, 0, dx, 0, sy, dy, 0, 0, 1];
    displayTestResult(canvas, checkMatrix(matrix, expectedValues));
  }

  public OnTestPerformance(canvas: drawing.Canvas) {
    let matrix = new drawing.Matrix();
    for (let i = 0; i < this.testCount_; i++) {
      matrix.postScale(2, 0.5, 0, 0);
    }
    displayTestResult(canvas, true);
  }
}

export class MatrixMapRect extends TestBase {
  public constructor() {
    super();
  }

  public OnTestFunction(canvas: drawing.Canvas) {
    let dst: common2D.Rect = { left: 100, top: 20, right: 130, bottom: 60 };
    let src: common2D.Rect = { left: 100, top: 80, right: 130, bottom: 120 };
    let brush = new drawing.Brush();
    let color: common2D.Color = { alpha: 255, red: 0, green: 255, blue: 0 };
    let matrix = new drawing.Matrix();
    canvas.setMatrix(matrix);
    brush.setColor(color);
    canvas.attachBrush(brush);
    console.log(TAG, "matrix.mapRect");
    if (matrix.mapRect(dst, src)) {
      console.log(TAG, "canvas.drawRect");
      canvas.drawRect(dst);
    }
    canvas.detachBrush();
  }

  public OnTestPerformance(canvas: drawing.Canvas) {
    let matrix = new drawing.Matrix();
    let src: common2D.Rect = { left: 100, top: 20, right: 130, bottom: 60 };
    let dst: common2D.Rect = { left: 100, top: 80, right: 130, bottom: 120 };
    for (let i = 0; i < this.testCount_; i++) {
      matrix.mapRect(dst, src);
    }
    displayTestResult(canvas, true);
  }
}

export class MatrixMapPoints extends TestBase {

  public constructor(){
    super();
  }

  public OnTestPerformance(canvas: drawing.Canvas) {

    let rand: OHRandom = new OHRandom();
    let src: Array<common2D.Point> = [];
    let pointsCount: number = 1000;
    for (let i = 0; i < pointsCount; i++) {
      src.push({x: rand.nextUScalar1() * 640, y: rand.nextUScalar1() * 480});
    }
    let matrix = new drawing.Matrix();

    for (let i = 0; i < this.testCount_; i++) {
      let dst: Array<common2D.Point> = matrix.mapPoints(src);
    }
    displayTestResult(canvas, true);
  }

  public OnTestFunction(canvas: drawing.Canvas) {
    {
      let matrix = new drawing.Matrix();
      canvas.setMatrix(matrix);

      let src: Array<common2D.Point> = [{x: 30, y:50}, {x: 10, y:15}, {x: 5, y:40}];
      let dst: Array<common2D.Point> = matrix.mapPoints(src);

      if (dst.length != 3) {
        console.error(TAG, 'Invalid dst length!');
        displayTestResult(canvas, false);
      } else {
        let pen = new drawing.Pen();
        this.ApplyPenStyle(pen, TestFunctionStyleType.DRAW_STYLE_TYPE_1);
        let brush = new drawing.Brush();
        this.ApplyBrushStyle (brush, TestFunctionStyleType.DRAW_STYLE_TYPE_1)
        canvas.attachPen(pen);
        canvas.attachBrush(brush);
        dst.forEach(point => {
          canvas.drawPoint(point.x, point.y)
        });
        canvas.detachPen();
        canvas.detachBrush();
      }
    }
  }
}
